File: y2023/d2/p2/cube_conundrum.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

type CubeSet struct {
	red   int
	green int
	blue  int
}

type Game struct {
	id      int
	subsets []CubeSet
}

func main() {
	if len(os.Args) < 1 {
		log.Panicln("invalid number of arguments passed, pass in the game information file path")
	}

	path := os.Args[1]
	games, err := parseAllGames(path)
	if err != nil {
		log.Panicf("error while reading game info, %s\n", err.Error())
	}

	sum := 0
	sumOfPower := 0
	for _, game := range games {
		if isGamePossible(game, 12, 13, 14) {
			sum += game.id
		}

		set := lowestPossibleSet(game)
		sumOfPower += (set.red * set.green * set.blue)
	}

	fmt.Println(sum)
	fmt.Println(sumOfPower)
}

func parseAllGames(path string) ([]*Game, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("calibration document is corrupt or does not exist, %w", err)
	}

	games := make([]*Game, 0)

	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanLines)
	for scanner.Scan() {
		if game, err := parseGame(scanner.Text()); err == nil {
			games = append(games, game)
		} else {
			return nil, err
		}
	}

	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("calibration document contents are invalid or file is corrupt, %w", err)
	}

	return games, nil
}

func parseGame(gameStr string) (*Game, error) {
	gameInfo := strings.Split(gameStr, ": ")
	id, err := strconv.Atoi(strings.Split(gameInfo[0], " ")[1])
	if err != nil {
		return nil, err
	}

	sets := make([]CubeSet, 0)
	for _, setStr := range strings.Split(gameInfo[1], "; ") {
		set := CubeSet{}
		for _, cubeStr := range strings.Split(setStr, ", ") {
			setInfo := strings.Split(cubeStr, " ")
			count, err := strconv.Atoi(setInfo[0])
			if err != nil {
				return nil, err
			}

			switch setInfo[1] {
			case "red":
				set.red = count
			case "blue":
				set.blue = count
			default:
				set.green = count
			}
		}
		sets = append(sets, set)
	}

	return &Game{id, sets}, nil
}

func isGamePossible(game *Game, maxRed, maxGreen, maxBlue int) bool {
	for _, set := range game.subsets {
		if set.red > maxRed || set.green > maxGreen || set.blue > maxBlue {
			return false
		}
	}

	return true
}

func lowestPossibleSet(game *Game) CubeSet {
	red := 0
	green := 0
	blue := 0

	for _, set := range game.subsets {
		if set.red > red {
			red = set.red
		}

		if set.green > green {
			green = set.green
		}

		if set.blue > blue {
			blue = set.blue
		}
	}

	return CubeSet{red, green, blue}
}

File: y2023/d2/p2/CubeConundrum.java
package main;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.Path;

public class CubeConundrum {

    public record CubeSet(int red, int green, int blue) {
    }

    public record Game(int id, CubeSet[] sets) {
    }

    public static void main(final String... args) throws IOException {
        if (args.length < 1)
            System.err.println("invalid number of arguments passed, pass in the game information file path");

        final var path = Paths.get(args[0]);
        final var games = parseAllGames(path);

        var sum = 0;
        var sumOfPower = 0;
        for (final var game : games) {
            if (isGamePossible(game, 12, 13, 14)) {
                sum += game.id;
            }

            final var set = lowestPossiblCubeSet(game);
            sumOfPower += (set.red * set.green * set.blue);
        }

        System.out.println(sum);
        System.out.println(sumOfPower);
    }

    private static Game[] parseAllGames(final Path path) throws IOException {
        return Files.lines(path)
                .map(CubeConundrum::parseGame)
                .toArray(Game[]::new);
    }

    private static Game parseGame(final String gameStr) {
        final var gameInfo = gameStr.split(": ");
        final var id = Integer.parseInt(gameInfo[0].substring(5));

        final var setsInfo = gameInfo[1].split("; ");
        final var sets = new CubeSet[setsInfo.length];
        var ptr = 0;
        for (final var set : setsInfo) {
            var red = 0;
            var green = 0;
            var blue = 0;

            for (final var cubeStr : set.split(", ")) {
                final var cubeInfo = cubeStr.split(" ");
                final var count = Integer.parseInt(cubeInfo[0]);
                switch (cubeInfo[1]) {
                    case "red":
                        red = count;
                        break;
                    case "blue":
                        blue = count;
                        break;
                    default:
                    case "green":
                        green = count;
                        break;
                }
            }

            sets[ptr++] = new CubeSet(red, green, blue);
        }

        return new Game(id, sets);
    }

    private static boolean isGamePossible(final Game game, final int maxRed, final int maxGreen, final int maxBlue) {
        for (final var set : game.sets) {
            if (maxRed < set.red || maxGreen < set.green || maxBlue < set.blue) {
                return false;
            }
        }

        return true;
    }

    private static CubeSet lowestPossiblCubeSet(final Game game) {
        var red = 0;
        var green = 0;
        var blue = 0;

        for (final var set : game.sets) {
            red = Math.max(red, set.red);
            green = Math.max(green, set.green);
            blue = Math.max(blue, set.blue);
        }

        return new CubeSet(red, green, blue);
    }

}
